### PR TITLE
[SPARK-39841][SQL] simplify conflict binary comparison

### PR DESCRIPTION
**Pull request by yikf**
### What changes were proposed in this pull request?
Currently, We do not handle conflicting binary comparison expressions, Let's say have a query, it filter `a < 0 and a > 1`, this condition should be a `FalseLiteral`, we can fold ahead of time to avoid unnecessary calculations.

Have a idea, we can capture a and conjunction and both it's left and right are `BinaryComparisonExpression`, we optimize this predicate as following step:

- Normal and left and right comparison expression, to keep Attribute is left side and literal is right side.
- If both comparison expression left child is semantic equal and right child is foldable, we compute a range for column predicate comparison .
- Then update this column comparison predicate by this range.

Before this pr:
image

After this pr:
image

### Why are the changes needed?
Fold ahead to avoid unnecessary calculations & improve performance

### Does this PR introduce _any_ user-facing change?
No, improve only

### How was this patch tested?
Add new tests

### CODE REVIEW

**Comment by yikf** [comment]
@cloud-fan could you please take a look when you have a time, thanks

**Comment by cloud-fan** [comment]
can you briefly explain your idea? Do you keep a range for each column and update the range when seeing a comparison? Then use the range to update the column comparison predicate?

**Comment by yikf** [comment]
> can you briefly explain your idea? Do you keep a range for each column and update the range when seeing a comparison? Then use the range to update the column comparison predicate?

Yea,  if capture a `and` conjunction and both it's `left` and `right` are `BinaryComparisonExpression`, we optimize this predicate as following step:

- Normal `and` left and right comparison expression, to keep `Attribute` is left side and `literal` is right side.
- If both comparison expression left child is semantic equal and right child is foldable, we compute  a range for comparison left side.
- Then update this column comparison predicate by this range.

**Comment by cloud-fan** [comment]
cc @gengliangwang

**Comment by yikf** [comment]
friendly ping @gengliangwang @cloud-fan , could you please take a look when you have a time?

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/SimplifyConflictBinaryComparisonSuite.scala [hunk 1]
```diff
@@ -0,0 +1,299 @@
...
+    //  a = 1 and a = 2  =====> FalseLiteral
+    val plan1 = nonNullRelation
+      .where(And(EqualTo($"a", Literal.apply(1)), EqualTo($"a", Literal.apply(2))))
+    val actual1 = Optimize.execute(plan1.analyze)
+    val correctAnswer1 = Optimize.execute(nonNullRelation.where(FalseLiteral).analyze)
+    comparePlans(actual1, correctAnswer1)
+
+    // a = 1 and a > 0  ====> a = 1
+    val plan2 = nonNullRelation
+      .where(And(EqualTo($"a", Literal.apply(1)), GreaterThan($"a", Literal.apply(0))))
```
Comment (gengliangwang) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/SimplifyConflictBinaryComparisonSuite.scala:53]: Shall we have test case where the literal is of different data type from the attribute?
For example, column `a` is an integer, and there is a predicate `a` = 1 AND a > 0.5

File: sql/core/src/test/scala/org/apache/spark/sql/jdbc/JDBCV2Suite.scala [hunk 2]
```diff
@@ -1939,7 +1939,7 @@ class JDBCV2Suite extends QueryTest with SharedSparkSession with ExplainSuiteHel
         "COUNT(CASE WHEN (SALARY > 8000.00) AND (SALARY <= 13000.00) THEN SALARY ELSE 0.00 END), " +
         "COUNT(CASE WHEN (SALARY > 11000.00) OR (SALARY < 10000.00) THEN SALARY ELSE 0.00 END), " +
         "COUNT(CASE WHEN (SALARY >= 12000.00) OR (SALARY < 9000.00) THEN SALARY ELSE 0.00 END), " +
-        "MAX(CASE WHEN (SALARY <= 10000.00) AND (SALARY >= 8000.00) THEN SALARY ELSE 0.00 END), " +
+        "MAX(CASE WHEN (SALARY >= 8000.00) AND (SALARY <= 10000.00) THEN SALARY ELSE 0.00 END), " +
```
Comment (gengliangwang) [inline sql/core/src/test/scala/org/apache/spark/sql/jdbc/JDBCV2Suite.scala:1942]: Why do we need to change this?

**Comment by gengliangwang** [comment]
cc @sigmod as well

**Comment by yikf** [comment]
friendly ping @gengliangwang @cloud-fan @sigmod, Sorry for late reply, i was busy last week, i fixed comments you left, please take a look again when you have a time~

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/SimplifyConflictBinaryComparisonSuite.scala (hunk 1 above)
Comment (yikf) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/SimplifyConflictBinaryComparisonSuite.scala:53]: updated

**Comment by wangyum** [comment]
@Yikf Is there any mainstream database that supports this optimization?

**Pull request closed without merging**